aten/src/ATen/mps/MPSAllocator.mm

```diff
@@ -292,9 +253,98 @@
   release_buffers(m_large_pool_shared);
   release_buffers(m_small_pool_private);
   release_buffers(m_small_pool_shared);
+  release_buffers(m_scalar_pool);
   return true;
 }
 
+// public interface to MPSAllocator
+id<MTLBuffer> MPSHeapAllocatorImpl::Malloc(size_t size, uint32_t usage)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+
+  BufferBlock* buffer_block = alloc_buffer_block(size, usage);
+  return buffer_block ? buffer_block->buffer : nullptr;
+}
+
+bool MPSHeapAllocatorImpl::isSharedBuffer(void* ptr)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+
+  BufferBlock *buffer_block = get_allocated_buffer_block(ptr);
+  // it's OK for the buffer_block to not exist yet
+  return buffer_block && (buffer_block->heap->pool->usage & UsageFlags::SHARED);
+}
+
+id<MTLBuffer> MPSHeapAllocatorImpl::getScalarBufferWithValue(void* value, size_t size, MPSStream* stream)
+{
+  BufferBlock* buffer_block = nullptr;
+  {
+    std::lock_guard<std::mutex> lock(m_mutex);
+
+    buffer_block = alloc_buffer_block(size, UsageFlags::SCALAR);
+    if (!buffer_block)
+      return nullptr;
+    memcpy([buffer_block->buffer contents], value, size);
+  }
+  // this will return the buffer block back to the pool after
+  // command buffer finishes copying scalar from CPU to GPU
+  [stream->commandBuffer() addCompletedHandler:^(id <MTLCommandBuffer>) {
+    //Free(buffer_block->buffer);
+    std::lock_guard<std::mutex> lock(m_mutex);
```

Comment:
https://developer.apple.com/documentation/metal/mtlcommandbuffer "When used, scheduled and completed handlers are blocks that are invoked in execution order. These handlers should perform quickly; if expensive or **blocking** work needs to be scheduled, defer that work to another thread." 
While I can't come up with any deadlock scenario off the top of my head with this code, I would recommend adding the `free_buffer` call to a serial dispatch queue.

Reply:
Actually, without this lock the code segfaults quickly as Metal calls this handler callback from a separate thread which causes a race condition while buffer is being re-inserted back into the pool. The dispatch queue won't work as the code depends on that mutex to synchronize the threads working on that pool. Creating a separate thread to 'defer' the work would be overkill just for this purpose. I can try os_unfair_lock instead of mutex, but don't expect any perf gain.